feat: Add support for design tokens (#34)
open-craft/edx-simple-theme: README.md

MFE_DEPLOY_NPM_OVERRIDES:
  - "@edx/brand@file:/edx/var/edxapp/themes/simple-theme/"
```

## WIP: Usage with style dictionary / design tokens

Design tokens present a new way of customising styles using variables that are
then translated into different forms depending on the platform they are used in.

> For instance a token like `color.brand.base` could translate to a SCSS variable
> called `$brand-base` a css variable called `--pgn-color-brand-base` or be output
> in an XML that can be used for styling Android. For now the focus of Paragon is
> on CSS variables.

Reviewer: ```suggestion
For instance a token like `color.brand.base` could translate to a SCSS variable
called `$brand-base`, a css variable called `--pgn-color-brand-base`, or be output
in an XML that can be used for styling Android. Currently, the focus of Paragon is on CSS variables.
```

To customise a theme using this system, you'll need to provide all the variables
that you want to override in JSON files. You can look at the JSON files in the
Paragon repository ([currently in the alpha branch](https://github.com/openedx/paragon/tree/alpha/tokens/src)) to see how they are used and what all can be
customised.

In the `tokens/core` or `tokens/light` directories (light is the theme name, you
can have other names) you can create JSON files that override any of the values
in Paragon. You can see how these JSON files are structured at the link above.
While Paragon follows a strucure above, based on component names, the variable
type etc. This is not necessary, the exact structure doesn't matter, they are
all loaded and merged into one dictionary of styles and the output theme will

> output it as CSS variables so they can be loaded by the custom theme.

Reviewer: We can probably add an [example](https://tasks.opencraft.com/browse/BB-8067?focusedCommentId=279648&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-279648) of the json file after this paragraph.

Author: I've linked to a folder filled with examples, and they can be quite verbose so, I think it doesn't need to be reproduced here.

This is a two-step process:

> 1. Building the tokens: The `paragon build-tokens` build tokens command compiles

Reviewer: ```suggestion
1. Building the tokens: The `npm build-tokens` build tokens command compiles
```

Author: The command being the scenes is `paragon`. I'm just running it via npm scripts.

Reviewer: @xitij2000 Yes, I know. People will need to use `npm` as they cannot directly run `paragon` commands which is why I am suggesting this change.

>    the JSON style dictionaries into CSS file sin the `paragon/css/` directory.

Reviewer: ```suggestion
   the JSON style dictionaries into CSS files in the `paragon/css/` directory.
```

   These are supposed to be committed to the repository.
2. Building the SCSS: The build token CSS files are then compiled into a brand

>    theme in the dist folder using the `paragon build-scss` command. The contents

Reviewer: ```suggestion
   theme in the dist folder using the `npm build-scss` command. The contents
```

MFE_DEPLOY_NPM_OVERRIDES:
  - "@edx/brand@file:/edx/var/edxapp/themes/simple-theme/"
```

> ## WIP: Usage with style dictionary / design tokens

Reviewer 2: What's left to clear this 'WIP' marker other than review?

Author: It was WIP since design tokens were WIP at the point when this was written. I will update it.

Design tokens present a new way of customising styles using variables that are
then translated into different forms depending on the platform they are used in.

For instance a token like `color.brand.base` could translate to a SCSS variable
called `$brand-base`, a css variable called `--pgn-color-brand-base`, or be output
in an XML that can be used for styling Android. Currently, the focus of Paragon is on CSS variables.

To customise a theme using this system, you'll need to provide all the variables
that you want to override in JSON files. You can look at the JSON files in the

> Paragon repository ([currently in the alpha branch](https://github.com/openedx/paragon/tree/alpha/tokens/src)) to see how they are used and what all can be

Reviewer 2: Design tokens landed in Teak so this is out of date, right?

          "urls": {
            "default": "https://cdn.jsdelivr.net/npm/@edx/paragon@$paragonVersion/dist/light.min.css",
            "brandOverride": "https://cdn.jsdelivr.net/npm/@edx/brand-edx.org@$brandVersion/dist/light.min.css"
          }
        }
      }
    }
  }
}

```

The above are the vanilla Paragon stylesheets. The core stylesheets contain the
bulk of the Paragon styling while the variants have the style variables that
modify how the core theme applies. For both of those you have a `default` and a
`brandOverride`. The default is the core theme whereas the `brandOverride` applies
on top. This keeps the theme smaller since you don't need to include and update
all the variables from the core theme.

In the above example when configuring the theme for an instance, you would
upload the `dist` folder contents to a CDN, and add the path to `core.min.css`
and `light.min.css` to the appropriate places. You can also create other themes
like a `dark` theme or a separate theme for each site.

NOTE: CSS variables are technically called CSS custom properties but they are
called variables here because that more closely corresponds with how they are
being used here.

> ## WIP: Automatically deploying a runtime theme to S3 via gitlab CI

Reviewer 2: Same question on the WIP here

This repo has a GitLab CI configuration to automatically deploy to AWS S3 if
configured.

> To use that you need to configure the environment variables that AWS CLI needs

Reviewer 2: This is cool, but I'm curious why we're adding it. Is it needed by one of our clients? Does it satisfy a workflow that the tutor plugin isn't handling? Or is this to make it more generally useful?

Author: Yes it was somewhat required for a client that had many themed sites so it was (at that point) easier to upload it to S3 and serve from there. With the latest changes that may no longer be required, but I'd need to test.
